packages/web3-eth/src/utils/detect_transaction_type.ts

```diff
@@ -17,26 +17,124 @@ along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
 import { TransactionTypeParser, Web3Context } from 'web3-core';
-import { EthExecutionAPI, Transaction, ETH_DATA_FORMAT } from 'web3-types';
-import { isNullish } from 'web3-validator';
+import { EthExecutionAPI, HardforksOrdered, Transaction, ETH_DATA_FORMAT } from 'web3-types';
+import { Web3ValidatorError, isNullish, validator } from 'web3-validator';
+import { InvalidPropertiesForTransactionTypeError } from 'web3-errors';
+
 import { InternalTransaction } from '../types';
 
+// undefined is treated as null for JSON schema validator
+const type0x0TransactionSchema = {
+	type: 'object',
+	properties: {
+		accessList: {
+			type: 'null',
+		},
+		maxFeePerGas: {
+			type: 'null',
+		},
+		maxPriorityFeePerGas: {
+			type: 'null',
+		},
+	},
+};
+const type0x1TransactionSchema = {
+	type: 'object',
+	properties: {
+		maxFeePerGas: {
+			type: 'null',
+		},
+		maxPriorityFeePerGas: {
+			type: 'null',
+		},
+	},
+};
+const type0x2TransactionSchema = {
+	type: 'object',
+	properties: {
+		gasPrice: {
+			type: 'null',
+		},
+	},
+};
+
+const validateTxTypeAndHandleErrors = (
+	txSchema: object,
+	tx: Transaction,
+	txType: '0x0' | '0x1' | '0x2',
+) => {
+	try {
+		validator.validateJSONSchema(txSchema, tx);
+	} catch (error) {
+		if (error instanceof Web3ValidatorError)
+			// Erroneously reported error
+			// eslint-disable-next-line @typescript-eslint/no-unsafe-call
+			throw new InvalidPropertiesForTransactionTypeError(error.errors, txType);
+
+		throw error;
+	}
+};
+
 export const defaultTransactionTypeParser: TransactionTypeParser = transaction => {
 	const tx = transaction as unknown as Transaction;
 
-	if (!isNullish(tx.type)) return format({ format: 'uint' }, tx.type, ETH_DATA_FORMAT);
+	if (!isNullish(tx.type)) {
+		let txSchema;
+		switch (tx.type) {
+			case '0x0':
+				txSchema = type0x0TransactionSchema;
+				break;
+			case '0x1':
+				txSchema = type0x1TransactionSchema;
+				break;
+			case '0x2':
+				txSchema = type0x2TransactionSchema;
+				break;
 
-	if (
-		!isNullish(tx.maxFeePerGas) ||
-		!isNullish(tx.maxPriorityFeePerGas) ||
-		tx.hardfork === 'london' ||
-		tx.common?.hardfork === 'london'
-	)
-		return '0x2';
+			default:
+				return format({ format: 'uint' }, tx.type, ETH_DATA_FORMAT);
+		}
+
+		validateTxTypeAndHandleErrors(txSchema, tx, tx.type);
 
-	if (!isNullish(tx.accessList) || tx.hardfork === 'berlin' || tx.common?.hardfork === 'berlin')
+		return format({ format: 'uint' }, tx.type, ETH_DATA_FORMAT);
+	}
+
+	// We don't check !isNullish(tx.gasPrice) here, because
+	// if it's not undefined, we still don't know if the network
+	// supports EIP-2718 (https://eips.ethereum.org/EIPS/eip-2718)
+	// and whether we should return undefined for legacy txs,
+	// or type 0x0 for legacy txs post EIP-2718
+
+	if (!isNullish(tx.accessList)) {
+		validateTxTypeAndHandleErrors(type0x1TransactionSchema, tx, '0x1');
 		return '0x1';
+	}
+
+	if (!isNullish(tx.maxFeePerGas) || !isNullish(tx.maxPriorityFeePerGas)) {
```

Comment:
if  `gasPrice` is given, there should be a case for that as well for returning `0x0 `and validating for `validateTxTypeAndHandleErrors(type0x1TransactionSchema, tx, '0x0');`

Reply:
or will it work with out that for 0x0, are there any test cases for it?

Reply:
Renamed [here](https://github.com/web3/web3.js/pull/6102/commits/cef42da04daf0db833f516c4bcfb7faf7b0884f9), also added a validation check for type `0x0` txs without `tx.type` specified

Reply:
@jdevcs The test you're describing is [here](https://github.com/web3/web3.js/blob/db3f3ee30c45025364d1f7106d2ae420ec8ec401/packages/web3-eth/test/fixtures/detect_transaction_type.ts#L202-L212) and the detected type would be `undefined`. Without `hardfork`, we can't be sure if the network supports EIP-2718, so we return `undefined` assuming is a pre-EIP-2718 legacy tx